File: app/fantasy420/src/fantasy420/app/Wrapped/wrapped.py
import concurrent.futures

import json
import requests
import sys

from bs4 import BeautifulSoup

year = 2022
league_ids = [203836968]

num_threads = 32

fetch_cache = {}

# for private leagues, need to store the cookie value of espn_s2 in espn_s2.txt
try:
    with open('espn_s2.txt') as fh:
        espn_s2 = fh.read().strip()
except:
    espn_s2 = None


class Vars:
    max_week_finished = 17


def main():
    wrapped = {}
    for league_id in league_ids:
        wrapped[league_id] = get_wrapped(league_id)
    print(json.dumps(wrapped))


def get_wrapped(league_id):
    team_names = get_team_names(league_id)
    weeks = get_weeks(league_id, len(team_names))
    populate_boxscores(weeks)
    populate_fieldgoals(weeks)
    players = get_players(weeks)
    return {"teamNames": team_names, "weeks": weeks, "players": players}


def fetch(url, decode_json=True):
    if url in fetch_cache:
        return fetch_cache[url]
    sys.stderr.write(url)
    raw_data = requests.get(url, cookies={'espn_s2': espn_s2})
    data = raw_data.json() if decode_json else raw_data.text
    fetch_cache[url] = data
    return data


def get_team_names(league_id):
    data = fetch(
        f"https://fantasy.espn.com/apis/v3/games/ffl/seasons/{year}/segments/0/leagues/{league_id}?view=mTeam"
    )
    return list(
        map(
            lambda team: f'{team["location"]} {team["nickname"]}',
            data["teams"],
        ), )


def get_weeks(league_id, num_teams):
    with concurrent.futures.ThreadPoolExecutor(num_threads) as executor:
        weeks = list(
            executor.map(
                lambda key: get_matches(league_id, key + 1, num_teams),
                list(range(Vars.max_week_finished)),
            ))
    weeks = [week for week in weeks if week]
    return weeks


def get_matches(league_id, week_num, num_teams):
    if week_num > Vars.max_week_finished:
        return None
    url = f"https://fantasy.espn.com/apis/v3/games/ffl/seasons/{year}/segments/0/leagues/{league_id}?view=mScoreboard&scoringPeriodId={week_num}"
    data = fetch(url)
    raw_matches = filter(
        lambda match: "rosterForCurrentScoringPeriod" in match["away"],
        data["schedule"],
    )
    matches = []
    for match in raw_matches:
        teams = list(map(
            get_team,
            [match["home"], match["away"]],
        ))
        matches.append(sorted(
            teams,
            key=lambda team: team["score"],
        ))
    if len(matches) != num_teams / 2 or matches[0][0]["score"] == 0:
        if Vars.max_week_finished > week_num:
            Vars.max_week_finished = week_num
        return None
    return {"number": week_num, "matches": matches}


def get_team(raw_team):
    pro_team_names = {
        0: 'FA',
        1: 'Atl',
        2: 'Buf',
        3: 'Chi',
        4: 'Cin',
        5: 'Cle',
        6: 'Dal',
        7: 'Den',
        8: 'Det',
        9: 'GB',
        10: 'Ten',
        11: 'Ind',
        12: 'KC',
        13: 'LV',
        14: 'LAR',
        15: 'Mia',
        16: 'Min',
        17: 'NE',
        18: 'NO',
        19: 'NYG',
        20: 'NYJ',
        21: 'Phi',
        22: 'Ari',
        23: 'Pit',
        24: 'LAC',
        25: 'SF',
        26: 'Sea',
        27: 'TB',
        28: 'Wsh',
        29: 'Car',
        30: 'Jax',
        33: 'Bal',
        34: 'Hou',
    }
    lineup = []
    for player in raw_team["rosterForMatchupPeriod"]["entries"]:
        scoring_players = list(
            filter(
                lambda scoring_player: scoring_player["playerId"] == player[
                    "playerId"],
                raw_team["rosterForCurrentScoringPeriod"]["entries"],
            ))
        if scoring_players and scoring_players[0]["lineupSlotId"] not in [
                20, 21
        ]:
            lineup.append(player["playerId"])
    roster = []
    for raw_player in raw_team["rosterForCurrentScoringPeriod"]["entries"]:
        player = raw_player["playerPoolEntry"]["player"]
        player_obj = {
            "id": player["id"],
            "name": player["fullName"],
            "score":
            get_points(raw_player["playerPoolEntry"]["appliedStatTotal"]),
            "position": player["defaultPositionId"],
            "team": pro_team_names[player["proTeamId"]].upper(),
        }
        if player_obj["name"] == "Rodrigo Blankenship":
            player_obj["team"] = "IND"
        roster.append(player_obj)
    score = get_points(
        sum([player["score"] for player in roster if player["id"] in lineup]))
    return {
        "teamIndex": raw_team["teamId"] - 1,
        "score": score,
        "lineup": [str(i) for i in lineup],
        "fullRoster": roster,
    }


def get_game_id(pro_team_name, week):
    url = f'https://www.espn.com/nfl/team/schedule/_/name/{pro_team_name}'
    schedule_html = fetch(url, decode_json=False)
    soup = BeautifulSoup(schedule_html, features="html.parser")
    t = soup.find("table")
    for row in t.findAll("tr"):
        if row.find("span") and row.find("span").text == str(week):
            hrefs = row.findAll("a", href=True)
            if len(hrefs) < 3:
                return None
            game_url = hrefs[2]["href"]
            return game_url.split("espn.com/nfl/game/_/gameId/")[-1]


def populate(weeks, week_key, f, position):
    to_fetch = []
    for week in weeks:
        for match in week["matches"]:
            for team in match:
                for player in team["fullRoster"]:
                    if player["position"] == position:
                        if player["team"] != "FA":
                            key = (player["team"], week["number"])
                            to_fetch.append(key)

    with concurrent.futures.ThreadPoolExecutor(num_threads) as executor:
        fetched_arr = list(executor.map(lambda g: f(*g), to_fetch))
    for week in weeks:
        week[week_key] = [
            i for i in [
                fetched_arr[i] for i, j in enumerate(to_fetch)
                if j[1] == week["number"]
            ] if i
        ]


def populate_boxscores(weeks):
    populate(weeks, "boxscores", get_boxscore, 16)  # DST = 16


def get_boxscore(pro_team_name, week):
    game_id = get_game_id(pro_team_name, week)
    if game_id is None: return None
    url = f'https://www.espn.com/nfl/boxscore/_/gameId/{game_id}'
    box_score_html = fetch(url, decode_json=False)
    __espnfitt__ = box_score_html.split("window['__espnfitt__']=")[-1].split(
        ";</script>\n")[0]
    parsed = json.loads(__espnfitt__)
    gamepackage = parsed["page"]["content"]["gamepackage"]

    bxscr = gamepackage["bxscr"]
    teamabbrevs = [i["tm"]["abbrev"] for i in bxscr]
    key = 1 - teamabbrevs.index(pro_team_name)

    stats = bxscr[key]["stats"]
    passing = float(
        next((s["ttls"][1] for s in stats if "passingYards" in s["keys"]),
             None))
    rushing = float(
        next((s["ttls"][1] for s in stats if "rushingYards" in s["keys"]),
             None))

    score = int(gamepackage["gmStrp"]["tms"][key]["score"])

    boxscore = {
        "team": teamabbrevs[1 - key],
        "oppTeam": pro_team_name,
        "passing": passing,
        "rushing": rushing,
        "score": score
    }

    return boxscore


def populate_fieldgoals(weeks):
    populate(weeks, "fieldgoals", get_fieldgoals, 5)  # K = 5


def get_fieldgoals(pro_team_name, week):
    game_id = get_game_id(pro_team_name, week)
    if game_id is None: return None
    url = f'https://www.espn.com/nfl/playbyplay/_/gameId/{game_id}'
    play_by_play_html = fetch(url, decode_json=False)
    soup = BeautifulSoup(play_by_play_html, features="html.parser")
    headline_divs = soup.findAll("div", class_="headline")
    headlines = [h.text for h in headline_divs if "Field Goal" in h.text]
    return {"team": pro_team_name, "fieldgoals": headlines}


def get_players(weeks):
    players = {}
    for week in weeks:
        for match in week["matches"]:
            for team in match:
                roster = {}
                for p in team["fullRoster"]:
                    players[p["id"]] = {
                        "name": p["name"],
                        "position": p["position"],
                        "team": p["team"]
                    }
                    roster[p["id"]] = p["score"]
                team["roster"] = roster
                del team["fullRoster"]
    return players


def get_points(raw_points):
    return round(raw_points, 2)


main()

File: extension/extensions/wrapped.js
function FetchWrapped() {
  const year = 2023;
  const leagueId =
    new URL(window.document.location.href).searchParams.get("leagueId") ||
    203836968;
  function fromEntries(arr) {
    return Object.fromEntries(
      arr.filter((a) => a !== undefined).map((a) => [a.key, a.value])
    );
  }
  function clog(t) {
    console.log(t);
    return t;
  }
  return Promise.resolve()
    .then(() => [
      Promise.resolve()
        .then(() =>
          fetch(
            `https://lm-api-reads.fantasy.espn.com/apis/v3/games/ffl/seasons/${year}/segments/0/leagues/${leagueId}?view=kona_playercard`,
            {
              headers: {
                accept: "application/json",
                "x-fantasy-filter": JSON.stringify({
                  players: {
                    filterStatsForTopScoringPeriodIds: {
                      value: 17,
                      additionalValue: [`00${year}`],
                    },
                  },
                }),
                "x-fantasy-platform":
                  "kona-PROD-5b4759b3e340d25d9e1ae248daac086ea7c37db7",
                "x-fantasy-source": "kona",
              },
              credentials: "include",
            }
          )
            .then((resp) => resp.json())
            .then((resp) =>
              resp.players
                .map((player) => player.player)
                .filter(
                  (player) => player.proTeamId !== 0 || player.onTeamId !== 0
                )
                .map((player) => ({
                  id: player.id.toString(),
                  nflTeamId: player.proTeamId.toString(),
                  name: player.fullName,
                  position:
                    { 1: "QB", 2: "RB", 3: "WR", 4: "TE", 5: "K", 16: "DST" }[
                      player.defaultPositionId
                    ] || player.defaultPositionId.toString(),
                  total: player.stats.find((stat) => stat.scoringPeriodId === 0)
                    .appliedTotal,
                  average: player.stats.find(
                    (stat) => stat.scoringPeriodId === 0
                  ).appliedAverage,
                  scores: fromEntries(
                    player.stats
                      .filter((stat) => stat.seasonId === year)
                      .map((stat) => ({
                        key: stat.scoringPeriodId.toString(),
                        value: parseFloat(stat.appliedTotal.toFixed(2)),
                      }))
                  ),
                }))
                .map((player) => ({ key: player.id, value: player }))
            )
            .then((playersArr) => fromEntries(playersArr))
        )
        .then((players) => players),
      Promise.resolve()
        .then(() =>
          fetch(
            `https://lm-api-reads.fantasy.espn.com/apis/v3/games/ffl/seasons/${year}/segments/0/leagues/${leagueId}?view=mRoster`
          )
            .then((resp) => resp.json())
            .then((resp) =>
              Promise.resolve()
                .then(() =>
                  Array.from(new Array(resp.status.latestScoringPeriod))
                    .map((_, i) => i + 1)
                    .map((weekNum) =>
                      fetch(
                        `https://lm-api-reads.fantasy.espn.com/apis/v3/games/ffl/seasons/${year}/segments/0/leagues/${leagueId}?view=mScoreboard&scoringPeriodId=${weekNum}`,
                        { credentials: "include" }
                      )
                        .then((resp) => resp.json())
                        .then((resp) =>
                          Promise.resolve()
                            .then(() =>
                              resp.teams.map((team) => ({
                                id: team.id.toString(),
                                name: team.name,
                                schedule: {
                                  weekNum,
                                  ...resp.schedule
                                    .flatMap((matchup) => [
                                      matchup.home,
                                      matchup.away,
                                    ])
                                    .find(
                                      (s) =>
                                        s.rosterForCurrentScoringPeriod &&
                                        s.teamId === team.id
                                    ),
                                },
                              }))
                            )
                            .then((week) =>
                              fromEntries(
                                week.map((team) => ({
                                  key: team.id,
                                  value: team,
                                }))
                              )
                            )
                        )
                    )
                )
                .then((ps) => Promise.all(ps))
                .then((weeks) =>
                  Object.values(weeks[0]).map((team) => ({
                    id: team.id,
                    name: team.name,
                    rosters: fromEntries(
                      weeks
                        .map((week) => week[team.id].schedule)
                        .filter((s) => s.rosterForCurrentScoringPeriod)
                        .map((s) => ({
                          weekNum: s.weekNum.toString(),
                          starting: s.rosterForMatchupPeriod.entries.map((e) =>
                            e.playerId.toString()
                          ),
                          rostered: s.rosterForCurrentScoringPeriod.entries.map(
                            (e) => e.playerId.toString()
                          ),
                        }))
                        .concat({
                          weekNum: "0",
                          starting: [],
                          rostered: resp.teams
                            .find((t) => t.id.toString() === team.id)
                            .roster.entries.map((e) => e.playerId.toString()),
                        })
                        .map((roster) => ({
                          key: roster.weekNum,
                          value: roster,
                        }))
                    ),
                  }))
                )
            )
            .then((teams) =>
              fromEntries(teams.map((team) => ({ key: team.id, value: team })))
            )
        )
        .then((teams) => teams),
      Promise.resolve()
        .then(() =>
          fetch(
            `https://lm-api-reads.fantasy.espn.com/apis/v3/games/ffl/seasons/${year}/segments/0/leagues/${leagueId}?view=mMatchupScore&view=mSettings`,
            { credentials: "include" }
          )
            .then((resp) => resp.json())
            .then((resp) =>
              Array.from(
                new Array(resp.settings.scheduleSettings.matchupPeriodCount)
              )
                .map((_, i) => i + 1)
                .map((matchupPeriodId) => ({
                  key: matchupPeriodId.toString(),
                  value: resp.schedule
                    .filter((s) => s.matchupPeriodId === matchupPeriodId)
                    .map((s) =>
                      [s.home, s.away].map((t) => t.teamId.toString())
                    ),
                }))
            )
            .then((matchups) => fromEntries(matchups))
        )
        .then((ffMatchups) => ffMatchups),
      Promise.resolve()
        .then(() =>
          fetch(
            `https://lm-api-reads.fantasy.espn.com/apis/v3/games/ffl/seasons/${year}?view=proTeamSchedules_wl`
          )
            .then((resp) => resp.json())
            .then((resp) =>
              resp.settings.proTeams.map((p) => ({
                id: p.id.toString(),
                name: p.name,
                byeWeek: p.byeWeek,
                proGamesByScoringPeriod: fromEntries(
                  Object.entries(p.proGamesByScoringPeriod)
                    .filter(([_, o]) => o[0].statsOfficial)
                    .map(([scoringPeriod, o]) => ({
                      key: scoringPeriod,
                      value: o[0].id,
                    }))
                ),
              }))
            )
            .then((nflTeams) =>
              Promise.resolve()
                .then(() =>
                  nflTeams.flatMap((team) =>
                    Object.values(team.proGamesByScoringPeriod)
                  )
                )
                .then((gameIds) =>
                  Object.keys(
                    fromEntries(
                      gameIds.map((gameId) => ({
                        key: gameId.toString(),
                        value: true,
                      }))
                    )
                  )
                )
                .then((gameIds) =>
                  gameIds.map((gameId) =>
                    fetch(
                      `https://www.espn.com/nfl/playbyplay/_/gameId/${gameId}`
                    )
                      .then((resp) => resp.text())
                      .then(
                        (resp) =>
                          resp
                            .split("window['__espnfitt__']=")
                            .reverse()[0]
                            .split(";</script>")[0]
                      )
                      .then((resp) => JSON.parse(resp))
                      .then((resp) =>
                        resp.page.content.gamepackage.scrSumm.scrPlayGrps
                          .flatMap((periodPlays) => periodPlays)
                          .filter((play) => play.typeAbbreviation === "FG")
                          .map((play) => ({
                            teamId: play.teamId,
                            yards: parseInt(
                              play.text
                                .split(" Yd Field Goal")[0]
                                .split(" ")
                                .reverse()[0]
                            ),
                          }))
                      )
                      .then((value) => ({ key: gameId.toString(), value }))
                  )
                )
                .then((ps) => Promise.all(ps))
                .then((gamesByGameId) => fromEntries(gamesByGameId))
                .then((gamesByGameId) =>
                  fetch(
                    `https://lm-api-reads.fantasy.espn.com/apis/v3/games/ffl/seasons/${year}/segments/0/leagues/${leagueId}?view=kona_playercard`,
                    {
                      headers: {
                        accept: "application/json",
                        "x-fantasy-filter": JSON.stringify({
                          players: {
                            filterSlotIds: { value: [16] },
                            filterStatsForTopScoringPeriodIds: { value: 17 },
                          },
                        }),
                        "x-fantasy-platform":
                          "kona-PROD-5b4759b3e340d25d9e1ae248daac086ea7c37db7",
                        "x-fantasy-source": "kona",
                      },
                      credentials: "include",
                    }
                  )
                    .then((resp) => resp.json())
                    .then((resp) =>
                      Object.keys(
                        fromEntries(
                          resp.players
                            .flatMap((player) => player.player.stats)
                            .map((s) => ({
                              key: s.scoringPeriodId.toString(),
                              value: true,
                            }))
                        )
                      ).map((scoringPeriodId) => ({
                        key: scoringPeriodId.toString(),
                        value: fromEntries(
                          resp.players
                            .map((player) => {
                              var _player$player$stats$;
                              return {
                                teamId: player.player.proTeamId,
                                stats:
                                  (_player$player$stats$ =
                                    player.player.stats.find(
                                      (s) =>
                                        s.scoringPeriodId.toString() ===
                                        scoringPeriodId
                                    )) === null ||
                                  _player$player$stats$ === void 0
                                    ? void 0
                                    : _player$player$stats$.stats,
                              };
                            })
                            .map(({ teamId, stats }) =>
                              stats === undefined
                                ? undefined
                                : {
                                    key: teamId.toString(),
                                    value: {
                                      yardsAllowed: stats["127"] || 0,
                                      pointsAllowed: stats["187"] || 0,
                                    },
                                  }
                            )
                        ),
                      }))
                    )
                    .then((defensesByScoringPeriod) =>
                      fromEntries(defensesByScoringPeriod)
                    )
                    .then((defensesByScoringPeriod) =>
                      nflTeams
                        .map(({ proGamesByScoringPeriod, ...team }) => ({
                          ...team,
                          nflGamesByScoringPeriod: fromEntries(
                            Object.entries(proGamesByScoringPeriod)
                              .filter(
                                ([scoringPeriod]) =>
                                  defensesByScoringPeriod[scoringPeriod] !==
                                  undefined
                              )
                              .map(([scoringPeriod, gameId]) => ({
                                key: scoringPeriod,
                                value: {
                                  fieldGoals: gamesByGameId[gameId]
                                    .filter((play) => play.teamId === team.id)
                                    .map((play) => play.yards),
                                  ...defensesByScoringPeriod[scoringPeriod][
                                    team.id
                                  ],
                                },
                              }))
                          ),
                        }))
                        .map((team) => ({ key: team.id, value: team }))
                    )
                )
                .then((nflTeams) => fromEntries(nflTeams))
            )
        )
        .then((nflTeams) => nflTeams),
    ])
    .then((ps) => Promise.all(ps))
    .then(([nflPlayers, ffTeams, ffMatchups, nflTeams]) => ({
      nflPlayers,
      ffTeams,
      ffMatchups,
      nflTeams,
    }))
    .then(console.log);
}
FetchWrapped();
